ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/common/logging/NamedLogging.scala

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.common.logging
+
+import java.util.concurrent.TimeUnit
+
+import com.typesafe.scalalogging.Logger
+import org.slf4j
+import org.slf4j.LoggerFactory
+
+trait NamedLogger {
+
+  /** Name for the logger. Can be empty. */
+  val name: String
+
+  def append(subName: String): NamedLogger
+
+  def forParticipant(id: String): NamedLogger = append(s"participant/$id")
+  def forDomain(id: String): NamedLogger = append(s"domain/$id")
+
+  private[logging] def createLogger(fullName: String): slf4j.Logger
+
+  /** create loggers in factory methods
+    *
+    * Sometimes, the NamedLogging trait can not be used, e.g. in a factory method. In these
+    * cases, a logger can be created using this function.
+    */
+  def createLogger(klass: Class[_]): com.typesafe.scalalogging.Logger = {
+    val fullName = Array(klass.getName, name)
+      .filterNot(_.isEmpty)
+      .mkString(":")
+    Logger(createLogger(fullName))
+  }
+
+}
+
+private[logging] case class SimpleNamedLogger(name: String) extends NamedLogger {
+  override def append(subName: String): NamedLogger =
+    if (name.isEmpty) SimpleNamedLogger(subName) else SimpleNamedLogger(s"$name/$subName")
+
+  override private[logging] def createLogger(fullName: String) = LoggerFactory.getLogger(fullName)
+}
+
+object NamedLogger {
+  def apply(name: String): NamedLogger = SimpleNamedLogger(name)
+  def apply(cls: Class[_]): NamedLogger = apply(cls.getSimpleName)
+
+  def forParticipant(name: String): NamedLogger = root.forParticipant(name)
+  def forDomain(name: String): NamedLogger = root.forDomain(name)
+  def root: NamedLogger = NamedLogger("")
+
+  def timedLog[T](hint: String, log: String => Unit)(exp: => T): T = {
+    log(s"TimedLog started [$hint]")
+    val started = System.nanoTime()
+    val ret = exp
+    val elapsed = System.nanoTime()
+    log(s"TimeLog finished [$hint] after ${TimeUnit.NANOSECONDS.toMillis(elapsed - started)} ms")
+    ret
+  }
+
+  def stackTrace(): String = {
+    Thread.currentThread().getStackTrace.take(2).map(_.toString).mkString("\n  ")
+  }
+
+}
+
+trait NamedLogging {
+  protected def namedLogger: NamedLogger
+
+  private[this] lazy val theLogger: Logger = namedLogger.createLogger(getClass)
```

Comment:
What's the reason for having `theLogger` and `logger`?

Reply:
fwiw, initially I simply copied our code as is, but still intended to add comments and adapt to what we need in the daml repo. It looks like this was meant to build the logger lazily - which may be a premature optimization. I guess I will get rid of theLogger.

Reply:
Removed